store/sqlstore/compliance_store.go

```diff
@@ -211,3 +211,109 @@ func (s SqlComplianceStore) ComplianceExport(job *model.Compliance) store.StoreC
+				Channels.Name AS ChannelName,
+				Channels.Header AS ChannelHeader,
+				Channels.Purpose AS ChannelPurpose,
+				Channels.LastPostAt AS ChannelLastPostAt,
+				Users.Id AS UserId,
+				Users.CreateAt AS UserCreateAt,
+				Users.UpdateAt AS UserUpdateAt,
+				Users.DeleteAt AS UserDeleteAt,
+				Users.Username AS UserUsername,
+				Users.Email AS UserEmail,
+				Users.Nickname AS UserNickname,
+				Users.FirstName AS UserFirstName,
+				Users.LastName AS UserLastName,
+				Teams.Id AS TeamId,
+				Teams.CreateAt AS TeamCreateAt,
+				Teams.UpdateAt AS TeamUpdateAt,
+				Teams.DeleteAt AS TeamDeleteAt,
+				Teams.DisplayName AS TeamDisplayName,
+				Teams.Name AS TeamName,
+				Teams.Description AS TeamDescription,
+				Teams.AllowOpenInvite AS TeamAllowOpenInvite, `
+
+		queryMySql :=
+			`CASE
+				WHEN Posts.Type = 'system_add_to_channel'
+				THEN (SELECT Email FROM Users WHERE Username = JSON_UNQUOTE(JSON_EXTRACT(CAST(Posts.Props AS JSON), '$.addedUsername')))
+				ELSE NULL END AS AddedUserEmail,
+			CASE
+				WHEN Posts.Type = 'system_remove_from_channel'
+				THEN (SELECT Email FROM Users WHERE Username = JSON_UNQUOTE(JSON_EXTRACT(CAST(Posts.Props AS JSON), '$.removedUsername')))
+				ELSE NULL END AS RemovedUserEmail `
+
+		queryPostgres :=
+			`CASE
+				WHEN Posts.Type = 'system_add_to_channel'
+				THEN (SELECT Email FROM Users WHERE Username = JSON_EXTRACT_PATH_TEXT(CAST(Posts.Props AS json), 'addedUsername'))
+				ELSE NULL END AS AddedUserEmail,
+			CASE
+				WHEN Posts.Type = 'system_remove_from_channel'
+				THEN (SELECT Email FROM Users WHERE Username = JSON_EXTRACT_PATH_TEXT(CAST(Posts.Props AS json), 'removedUsername'))
+				ELSE NULL END AS RemovedUserEmail `
+
+		querySuffix :=
+			`FROM
+				Posts
+				LEFT OUTER JOIN Channels ON Posts.ChannelId = Channels.Id
+				LEFT OUTER JOIN Users ON Posts.UserId = Users.Id
+				LEFT OUTER JOIN Teams ON Channels.TeamId = Teams.Id
+			WHERE
+				Posts.CreateAt > :StartTime
+			ORDER BY PostCreateAt
+			LIMIT :Limit`
+
+		var cposts []*model.MessageExport
+		result := store.StoreResult{}
+
+		query := queryPrefix
+		if s.DriverName() != model.DATABASE_DRIVER_MYSQL && s.DriverName() != model.DATABASE_DRIVER_POSTGRES {
+			result.Err = model.NewAppError("SqlPreferenceStore.save", "store.sql_preference.save.missing_driver.app_error", nil, "Failed to update preference because of missing driver", http.StatusNotImplemented)
+		} else {
+			if s.DriverName() == model.DATABASE_DRIVER_MYSQL {
+				query += queryMySql
+			} else if s.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+				query += queryPostgres
+			}
+			query += querySuffix
+
+			if _, err := s.GetReplica().Select(&cposts, query, props); err != nil {
+				result.Err = model.NewAppError("SqlComplianceStore.MessageExport", "store.sql_compliance.message_export.app_error", nil, err.Error(), http.StatusInternalServerError)
+			} else {
+				result.Data = cposts
+			}
+		}
+
+		storeChannel <- result
+		close(storeChannel)
+	}()
```

Comment:
There a reason you can't simplify this with `store.Do`?

Reply:
Nope. The code happens to be older than `store.Do` though. One of the many issues caused by maintaining a long-running branch. I don't get the benefits of all of your refactors.